File: packages/process-extract/src/extract-process.ts
import {
  err,
  ok,
  type PipelineContext,
  type Process,
  type ProcessError,
  proc,
  type Result,
  type RetryPolicy,
} from '@idriszade/core';
import type { ZodType } from 'zod';
import { zodToJsonSchema } from 'zod-to-json-schema';
import { withExtractSpan } from './otel.js';
import type { ProviderCallParams, ProviderResponse } from './providers/types.js';

export interface ExtractProcessConfig<I, O> {
  id?: string;
  provider: 'openai' | 'anthropic' | 'gemini';
  model: string;
  prompt: string | ((input: I) => string);
  inputSchema?: ZodType<I>;
  outputSchema: ZodType<O>;
  apiKey?: string;
  temperature?: number;
  maxRetriesOnSchemaFailure?: number;
  systemPrompt?: string;
  retryPolicy?: Partial<RetryPolicy>;
}

export type ExtractError = ProcessError;

function resolveApiKey(provider: string, configKey: string | undefined): string | undefined {
  if (configKey) return configKey;
  switch (provider) {
    case 'openai':
      return process.env.OPENAI_API_KEY;
    case 'anthropic':
      return process.env.ANTHROPIC_API_KEY;
    case 'gemini':
      return process.env.GEMINI_API_KEY ?? process.env.GOOGLE_API_KEY;
    default:
      return undefined;
  }
}

function mapProviderError(e: unknown, provider: string): ProcessError {
  if (!(e instanceof Error)) {
    return {
      type: 'unknown',
      code: 'unknown_error',
      message: String(e),
    };
  }

  const msg = e.message;
  const status = (e as { status?: number }).status;
  const code = (e as { code?: string }).code;

  // Content filter (OpenAI null content)
  if (code === 'content_filter') {
    return { type: 'permanent', code: 'content_filtered', message: msg };
  }

  // Provider not installed
  if (
    msg.includes('Cannot find module') ||
    msg.includes('Module not found') ||
    msg.includes('ERR_MODULE_NOT_FOUND') ||
    msg.includes('Failed to resolve') ||
    msg.includes('Could not resolve')
  ) {
    const sdkMap: Record<string, string> = {
      openai: 'openai',
      anthropic: '@anthropic-ai/sdk',
      gemini: '@google/generative-ai',
    };
    const sdk = sdkMap[provider] ?? provider;
    return {
      type: 'permanent',
      code: 'provider_not_installed',
      message: `Install the SDK: npm i ${sdk}`,
      doc_url: '/docs/install-providers.md',
    };
  }

  // Rate limit
  if (status === 429 || msg.includes('rate') || msg.includes('429') || msg.includes('quota')) {
    const retryAfterMatch = /retry.after[:\s]+(\d+)/i.exec(msg);
    const retry_after_ms =
      retryAfterMatch !== null && retryAfterMatch[1] !== undefined
        ? parseInt(retryAfterMatch[1], 10) * 1000
        : undefined;
    return {
      type: 'transient',
      code: 'rate_limited',
      message: msg,
      ...(retry_after_ms !== undefined ? { metadata: { retry_after_ms } } : {}),
    };
  }

  // Auth failure
  if (
    status === 401 ||
    msg.includes('auth') ||
    msg.includes('API key') ||
    msg.includes('Unauthorized')
  ) {
    return {
      type: 'permanent',
      code: 'auth_failed',
      message: msg,
    };
  }

  // Network / timeout
  if (
    msg.includes('network') ||
    msg.includes('ECONNREFUSED') ||
    msg.includes('ETIMEDOUT') ||
    msg.includes('fetch failed') ||
    msg.includes('timeout')
  ) {
    return {
      type: 'transient',
      code: 'network_error',
      message: msg,
    };
  }

  // Default fallback: 4xx → permanent, everything else → unknown
  if (status !== undefined && status >= 400 && status < 500) {
    return { type: 'permanent', code: 'provider_error', message: msg };
  }
  return { type: 'unknown', code: 'provider_unknown', message: msg };
}

async function callProvider(
  provider: 'openai' | 'anthropic' | 'gemini',
  params: ProviderCallParams,
): Promise<ProviderResponse> {
  switch (provider) {
    case 'openai': {
      const { callOpenAI } = await import('./providers/openai.js');
      return callOpenAI(params);
    }
    case 'anthropic': {
      const { callAnthropic } = await import('./providers/anthropic.js');
      return callAnthropic(params);
    }
    case 'gemini': {
      const { callGemini } = await import('./providers/gemini.js');
      return callGemini(params);
    }
  }
}

export function createExtractProcess<I, O>(config: ExtractProcessConfig<I, O>): Process<I, O> {
  const resolvedId = config.id ?? proc();
  const maxRetries = config.maxRetriesOnSchemaFailure ?? 2;

  async function run(input: I, _ctx: PipelineContext): Promise<Result<O, ProcessError>> {
    const promptStr = typeof config.prompt === 'function' ? config.prompt(input) : config.prompt;

    // zod-to-json-schema is pinned to Zod v3 types; cast through unknown to satisfy its ZodSchema<any> param
    const jsonSchemaRaw = zodToJsonSchema(config.outputSchema as never, { $refStrategy: 'none' });
    const jsonSchema = jsonSchemaRaw as Record<string, unknown>;

    const apiKey = resolveApiKey(config.provider, config.apiKey);

    const baseParams: ProviderCallParams = {
      model: config.model,
      apiKey,
      systemPrompt: config.systemPrompt,
      promptStr,
      jsonSchema,
      temperature: config.temperature,
    };

    let lastParseError = '';
    let lastIssues: unknown[] = [];

    for (let attempt = 0; attempt <= maxRetries; attempt++) {
      let providerResponse: ProviderResponse;

      const callParams: ProviderCallParams =
        attempt === 0
          ? baseParams
          : {
              ...baseParams,
              promptStr: `${baseParams.promptStr}\n\nYour last response failed schema validation: ${lastParseError}. Output valid JSON matching: ${JSON.stringify(jsonSchema)}`,
            };

      try {
        providerResponse = await withExtractSpan(
          config.provider,
          config.model,
          async (setUsage) => {
            const resp = await callProvider(config.provider, callParams);
            setUsage(resp.inputTokens, resp.outputTokens, resp.finishReasons);
            return resp;
          },
        );
      } catch (e) {
        return err(mapProviderError(e, config.provider));
      }

      const parsed = config.outputSchema.safeParse(providerResponse.rawJson);
      if (parsed.success) {
        return ok(parsed.data);
      }

      lastParseError = parsed.error.message;
      lastIssues = parsed.error.issues;
    }

    return err({
      type: 'validation',
      code: 'schema_parse_exhausted',
      message: `Schema validation failed after ${maxRetries + 1} attempts: ${lastParseError}`,
      metadata: { issues: lastIssues },
    });
  }

  return {
    id: resolvedId,
    inputSchema: config.inputSchema,
    outputSchema: config.outputSchema,
    retryPolicy: config.retryPolicy,
    run,
  };
}

File: packages/adapter-python-process-extract/src/pkit_process_extract/extract_process.py
"""extract_process — Pydantic-based port of process-extract TS adapter.

Mirrors @idriszade/process-extract: LLM-powered structured data extraction
with schema validation and retry-on-failure loop.

ADR IX-1 wire protocol is NOT used here — this is a direct Python library
(not a subprocess adapter). The caller invokes .run(input) directly.
"""

from __future__ import annotations

import json
import os
from dataclasses import dataclass, field
from typing import Any, Generic, Literal, Type, TypeVar

from pydantic import BaseModel, ValidationError

_O = TypeVar("_O", bound=BaseModel)


class ExtractProcessConfig(BaseModel, Generic[_O]):
    """Configuration for create_extract_process.

    Mirrors ExtractProcessConfig<I, O> from TS but simplified for Python:
    - provider names match TS ("anthropic" | "openai" | "google" vs "gemini" in TS)
    - prompt is a plain string; pass {input} placeholder to inject the input value
    - output_schema is a Pydantic model class (not Zod type)
    """

    provider: Literal["anthropic", "openai", "google"]
    model: str
    prompt: str
    output_schema: Type[BaseModel]
    api_key: str | None = None
    temperature: float = 0.0
    max_retries_on_schema_failure: int = 2
    system_prompt: str | None = None

    model_config = {"arbitrary_types_allowed": True}


@dataclass
class ProcessError:
    """Error envelope returned when extraction fails."""

    code: str  # "schema_validation_failed" | "provider_not_installed" | "api_error" | ...
    message: str
    cause: Exception | None = field(default=None, repr=False)


@dataclass
class ProcessResult(Generic[_O]):
    """Result discriminant: ok=True means value is set; ok=False means error is set."""

    ok: bool
    value: _O | None = None
    error: ProcessError | None = None


def _resolve_api_key(provider: str, config_key: str | None) -> str | None:
    """Fall back to env vars if no explicit key provided (mirrors TS resolveApiKey)."""
    if config_key:
        return config_key
    env_map = {
        "anthropic": "ANTHROPIC_API_KEY",
        "openai": "OPENAI_API_KEY",
        "google": "GEMINI_API_KEY",
    }
    env_var = env_map.get(provider)
    if env_var:
        return os.environ.get(env_var) or os.environ.get("GOOGLE_API_KEY")
    return None


def _import_anthropic() -> Any:
    """Lazy import Anthropic SDK with a helpful error if not installed."""
    try:
        from anthropic import Anthropic  # type: ignore[import-untyped]

        return Anthropic
    except ImportError as exc:
        raise RuntimeError(
            "anthropic package not installed. "
            "pip install pkit-process-extract[anthropic]"
        ) from exc


def _import_openai() -> Any:
    """Lazy import OpenAI SDK with a helpful error if not installed."""
    try:
        from openai import OpenAI  # type: ignore[import-untyped]

        return OpenAI
    except ImportError as exc:
        raise RuntimeError(
            "openai package not installed. "
            "pip install pkit-process-extract[openai]"
        ) from exc


def _import_google() -> Any:
    """Lazy import Google Generative AI SDK with a helpful error if not installed."""
    try:
        import google.generativeai as genai  # type: ignore[import-untyped]

        return genai
    except ImportError as exc:
        raise RuntimeError(
            "google-generativeai package not installed. "
            "pip install pkit-process-extract[google]"
        ) from exc


def _call_anthropic(
    model: str,
    api_key: str | None,
    system_prompt: str | None,
    prompt_str: str,
    json_schema: dict[str, Any],
    temperature: float,
) -> Any:
    """Call Anthropic API and return parsed JSON dict."""
    Anthropic = _import_anthropic()
    client = Anthropic(api_key=api_key) if api_key else Anthropic()

    messages: list[dict[str, Any]] = [{"role": "user", "content": prompt_str}]

    kwargs: dict[str, Any] = {
        "model": model,
        "max_tokens": 4096,
        "messages": messages,
        "tools": [
            {
                "name": "extract_data",
                "description": "Extract structured data matching the schema",
                "input_schema": json_schema,
            }
        ],
        "tool_choice": {"type": "tool", "name": "extract_data"},
    }
    if system_prompt:
        kwargs["system"] = system_prompt

    response = client.messages.create(**kwargs)
    # Find tool use block
    for block in response.content:
        if hasattr(block, "type") and block.type == "tool_use":
            return block.input
    raise ValueError("Anthropic response did not contain tool_use block")


def _call_openai(
    model: str,
    api_key: str | None,
    system_prompt: str | None,
    prompt_str: str,
    json_schema: dict[str, Any],
    temperature: float,
) -> Any:
    """Call OpenAI API and return parsed JSON dict."""
    OpenAI = _import_openai()
    client = OpenAI(api_key=api_key) if api_key else OpenAI()

    messages: list[dict[str, Any]] = []
    if system_prompt:
        messages.append({"role": "system", "content": system_prompt})
    messages.append({"role": "user", "content": prompt_str})

    response = client.chat.completions.create(
        model=model,
        messages=messages,
        temperature=temperature,
        response_format={
            "type": "json_schema",
            "json_schema": {"name": "extract_output", "schema": json_schema, "strict": True},
        },
    )
    content = response.choices[0].message.content
    if content is None:
        raise ValueError("OpenAI returned null content (likely content filter)")
    return json.loads(content)


def _call_google(
    model: str,
    api_key: str | None,
    system_prompt: str | None,
    prompt_str: str,
    json_schema: dict[str, Any],
    temperature: float,
) -> Any:
    """Call Google Generative AI API and return parsed JSON dict."""
    genai = _import_google()

    if api_key:
        genai.configure(api_key=api_key)

    generation_config: dict[str, Any] = {
        "temperature": temperature,
        "response_mime_type": "application/json",
        "response_schema": json_schema,
    }

    model_kwargs: dict[str, Any] = {}
    if system_prompt:
        model_kwargs["system_instruction"] = system_prompt

    g_model = genai.GenerativeModel(model, **model_kwargs)
    response = g_model.generate_content(prompt_str, generation_config=generation_config)
    return json.loads(response.text)


def _map_provider_error(exc: Exception, provider: str) -> ProcessError:
    """Map a provider exception to a ProcessError with actionable code."""
    msg = str(exc)
    status: int | None = getattr(exc, "status_code", getattr(exc, "status", None))

    if "not installed" in msg and "pip install" in msg:
        return ProcessError(code="provider_not_installed", message=msg, cause=exc)

    if status == 401 or any(k in msg.lower() for k in ("auth", "api key", "unauthorized")):
        return ProcessError(code="auth_failed", message=msg, cause=exc)

    if status == 429 or any(k in msg.lower() for k in ("rate", "quota", "429")):
        return ProcessError(code="rate_limited", message=msg, cause=exc)

    if any(k in msg.lower() for k in ("network", "econnrefused", "etimedout", "timeout")):
        return ProcessError(code="network_error", message=msg, cause=exc)

    if status is not None and 400 <= status < 500:
        return ProcessError(code="api_error", message=msg, cause=exc)

    return ProcessError(code="api_error", message=msg, cause=exc)


class _ExtractProcess(Generic[_O]):
    """Callable process object returned by create_extract_process."""

    def __init__(self, config: ExtractProcessConfig[_O]) -> None:
        self._config = config
        self._max_retries = config.max_retries_on_schema_failure

    async def run(self, input_value: Any) -> ProcessResult[_O]:
        """Run extraction against the configured provider, retrying on schema failure."""
        config = self._config
        prompt_str = config.prompt.replace("{input}", str(input_value))
        json_schema = config.output_schema.model_json_schema()
        api_key = _resolve_api_key(config.provider, config.api_key)

        last_parse_error = ""

        for attempt in range(self._max_retries + 1):
            if attempt > 0:
                prompt_str = (
                    f"{config.prompt.replace('{input}', str(input_value))}\n\n"
                    f"Your last response failed schema validation: {last_parse_error}. "
                    f"Output valid JSON matching: {json.dumps(json_schema)}"
                )

            try:
                raw_data = self._call_provider(
                    api_key=api_key,
                    prompt_str=prompt_str,
                    json_schema=json_schema,
                )
            except RuntimeError as exc:
                # Provider not installed
                return ProcessResult(
                    ok=False,
                    error=ProcessError(
                        code="provider_not_installed",
                        message=str(exc),
                        cause=exc,
                    ),
                )
            except Exception as exc:
                return ProcessResult(
                    ok=False,
                    error=_map_provider_error(exc, config.provider),
                )

            try:
                validated = config.output_schema.model_validate(raw_data)
                return ProcessResult(ok=True, value=validated)
            except ValidationError as exc:
                last_parse_error = str(exc)

        return ProcessResult(
            ok=False,
            error=ProcessError(
                code="schema_validation_failed",
                message=(
                    f"Schema validation failed after {self._max_retries + 1} attempts: "
                    f"{last_parse_error}"
                ),
            ),
        )

    def _call_provider(
        self,
        api_key: str | None,
        prompt_str: str,
        json_schema: dict[str, Any],
    ) -> Any:
        config = self._config
        if config.provider == "anthropic":
            return _call_anthropic(
                model=config.model,
                api_key=api_key,
                system_prompt=config.system_prompt,
                prompt_str=prompt_str,
                json_schema=json_schema,
                temperature=config.temperature,
            )
        if config.provider == "openai":
            return _call_openai(
                model=config.model,
                api_key=api_key,
                system_prompt=config.system_prompt,
                prompt_str=prompt_str,
                json_schema=json_schema,
                temperature=config.temperature,
            )
        if config.provider == "google":
            return _call_google(
                model=config.model,
                api_key=api_key,
                system_prompt=config.system_prompt,
                prompt_str=prompt_str,
                json_schema=json_schema,
                temperature=config.temperature,
            )
        raise ValueError(f"Unknown provider: {config.provider!r}")


def create_extract_process(config: ExtractProcessConfig[_O]) -> _ExtractProcess[_O]:
    """Factory matching TS createExtractProcess.

    Returns an object with async .run(input) -> ProcessResult[_O].
    """
    return _ExtractProcess(config)
